### PR TITLE
Documentation, SQLExec and SQLQuery commands

### PR DESCRIPTION
**Pull request by raducrisan1**

### CODE REVIEW

File: ImmuDB4Net.Tests/BasicClientIntegrationTests.cs [hunk 1]
```diff
@@ -132,6 +132,25 @@ public async Task Test2()
...
+        Assert.AreEqual(1, rspCreate.Items.Count);
+        var rspIndex = await client.SQLExec("CREATE INDEX IF NOT EXISTS ON logs(created)");
+        Assert.AreEqual(1, rspIndex.Items.Count);
+        var rspInsert = await client.SQLExec("INSERT INTO logs(created, entry) VALUES($1, $2)", 
+            new SQL.SQLParameter(DateTime.UtcNow),
+            new SQL.SQLParameter("entry1"));
+        var queryResult = await client.SQLQuery("SELECT created, entry FROM LOGS order by created DESC");
+        Assert.AreEqual(2, queryResult.Columns.Count);
+        Assert.AreEqual(1, queryResult.Rows.Count);
+        var sqlVal = queryResult.Rows[0]["(defaultdb.logs.entry)"];
```
Comment (jeroiraz) [inline ImmuDB4Net.Tests/BasicClientIntegrationTests.cs:152]: queryResult object could provide a higher level of abstraction to consume returned row data.
the column encoding `(defaultdb.logs.entry)` could be hidden and it might be possible to read the value just by using the column name

File: ImmuDB4Net/ImmuClient.cs [hunk 2]
```diff
@@ -1363,4 +1365,117 @@ private List<Tx> buildList(ImmudbProxy.TxList txList)
...
+        var req = new ImmudbProxy.SQLQueryRequest
+        {
+            Sql = sqlStatement,
+        };
+        if (parameters != null)
+        {
+            int paramNameCounter = 1;
+            foreach (var entry in parameters)
+            {
+               var namedParam = new NamedParam
```
Comment (jeroiraz) [inline ImmuDB4Net/ImmuClient.cs:1459]: may need to validate entry is not null

File: ImmuDB4Net.Tests/BasicClientIntegrationTests.cs (hunk 1 above)
Comment (raducrisan1) [reply ImmuDB4Net.Tests/BasicClientIntegrationTests.cs]: done

File: ImmuDB4Net/ImmuClient.cs [hunk 3]
```diff
@@ -1363,4 +1365,117 @@ private List<Tx> buildList(ImmudbProxy.TxList txList)
...
+        var req = new ImmudbProxy.SQLExecRequest
+        {
+            Sql = sqlStatement,
+        };
+        if (parameters != null)
+        {
+            int paramNameCounter = 1;
+            foreach (var entry in parameters)
+            {
+                var namedParam = new NamedParam
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1445]: done, added support for null value as well

File: ImmuDB4Net/ImmuClient.cs [hunk 4]
```diff
@@ -1363,4 +1365,117 @@ private List<Tx> buildList(ImmudbProxy.TxList txList)
...
+                };
+                req.Params.Add(namedParam);
+            }
+        }
+        var result = await Service.SQLExecAsync(req, Service.GetHeaders(ActiveSession));
+        var sqlResult = new SQL.SQLExecResult();
+        
+        foreach (var item in result.Txs)
+        {
+            sqlResult.Items.Add(new SQLExecResultItem { TxID = item.Header.BlTxId, UpdatedRowsCount = item.UpdatedRows });
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs]: done

File: ImmuDB4Net/ImmuClient.cs (hunk 2 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs]: done

File: ImmuDB4Net/ImmuClient.cs [hunk 5]
```diff
@@ -94,37 +90,99 @@ internal IConnection Connection
...
+    /// <param name="serverPort">The port where the ImmuDB server listens</param>
     public ImmuClient(string serverUrl, int serverPort)
         : this(NewBuilder().WithServerUrl(serverUrl).WithServerPort(serverPort))
     {
     }
 
+    /// <summary>
+    /// Initializes a new instance of <see cref="ImmuClient" />
+    /// </summary>
+    /// <param name="serverUrl">The ImmuDB server address, ex: localhost or http://localhost </param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:183]:
```suggestion
    /// <param name="serverUrl">The ImmuDB server address, e.g. localhost or http://localhost </param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 6]
```diff
@@ -94,37 +90,99 @@ internal IConnection Connection
...
+    /// <returns></returns>
     public ImmuClient() : this(NewBuilder())
     {
 
     }
 
+    /// <summary>
+    ///  Initializes a new instance of <see cref="ImmuClient" />. It uses the implicit 'defaultdb' value for the database name.
+    /// </summary>
+    /// <param name="serverUrl">The ImmuDB server address, ex: localhost or http://localhost </param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:173]:
```suggestion
    /// <param name="serverUrl">The ImmuDB server address, e.g. localhost or http://localhost </param>
```

(http://www.jvhighschool.com/study-tip-how-to-use-i-e-e-g-and-ex-properly/#:~:text=E.g.%20is%20another%20Latin%20abbreviations,actually%20stands%20for%20“exercise.”)

File: ImmuDB4Net/ImmuClient.cs [hunk 7]
```diff
@@ -136,17 +194,20 @@ internal ImmuClient(ImmuClientBuilder builder)
...
         heartbeatInterval = builder.HeartbeatInterval;
         ConnectionShutdownTimeout = builder.ConnectionShutdownTimeout;
         stateHolder.DeploymentInfoCheck = builder.DeploymentInfoCheck;
         stateHolder.DeploymentKey = Utils.GenerateShortHash(GrpcAddress);
         stateHolder.DeploymentLabel = GrpcAddress;
     }
 
-
+    /// <summary>
+    /// Releases the resources used by the SDK objects. (ex: connection pool resources). As best practice, this method should be call just before the existing process ends.
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:208]:
```suggestion
    /// Releases the resources used by the SDK objects. (e.g. connection pool resources). As best practice, this method should be call just before the existing process ends.
```

File: ImmuDB4Net/ImmuClient.cs [hunk 8]
```diff
@@ -94,37 +90,99 @@ internal IConnection Connection
...
         : this(NewBuilder().WithServerUrl(serverUrl).WithServerPort(serverPort))
     {
     }
 
+    /// <summary>
+    /// Initializes a new instance of <see cref="ImmuClient" />
+    /// </summary>
+    /// <param name="serverUrl">The ImmuDB server address, ex: localhost or http://localhost </param>
+    /// <param name="serverPort">The port where the ImmuDB server listens</param>
+    /// <param name="database">The name of the ImmuDB server's database to use</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:185]: Just noticed that - what's the use case for this constructor? We will have to specify the database name in `Open` anyway?

File: ImmuDB4Net/ImmuClient.cs [hunk 9]
```diff
@@ -217,6 +334,11 @@ public async Task Open(string username, string password, string defaultdb)
         }
     }
 
+    /// <summary>
+    /// Releases to the connection pool the established connection and acquires a new one.
+    /// </summary>
+    /// <remarks>The active session not affected.</remarks>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:340]:
```suggestion
    /// <remarks>The active session is not affected.</remarks>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 10]
```diff
@@ -217,6 +334,11 @@ public async Task Open(string username, string password, string defaultdb)
         }
     }
 
+    /// <summary>
+    /// Releases to the connection pool the established connection and acquires a new one.
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:338]:
```suggestion
    /// Releases the established connection to the connection pool and acquires a new one.
```

File: ImmuDB4Net/ImmuClient.cs [hunk 11]
```diff
@@ -276,45 +403,59 @@ private void CheckSessionHasBeenOpened()
 
     private object stateSync = new Object();
 
-    public ImmuState State()
+    /// <summary>
+    /// Gets the database state data and if not present then it is updated from the server
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:407]:
```suggestion
    /// Gets the database state data and if not present updates it from the server
```

File: ImmuDB4Net/ImmuClient.cs [hunk 12]
```diff
@@ -337,6 +483,10 @@ public async Task UseDatabase(string database)
         currentDb = database;
     }
 
+    /// <summary>
+    /// Gets the server's database list
+    /// </summary>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:488]: Please add a remark that this will only return databases the user has access to.

File: ImmuDB4Net/ImmuClient.cs [hunk 13]
```diff
@@ -354,6 +504,12 @@ public async Task<List<string>> Databases()
     // ========== GET ==========
     //
 
+    /// <summary>
+    /// Retrieves the value for a specific key at a transaction ID.
+    /// </summary>
+    /// <param name="key">The lookup key</param>
+    /// <param name="tx">The transaction id</param>
+    /// <returns>An <see cref="Entry"/> object. Most often the Value field is used.</returns>
     public async Task<Entry> GetAtTx(byte[] key, ulong tx)
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:513]: Why this is `GetAtTx` where other methods overload `Get` ?

File: ImmuDB4Net/ImmuClient.cs [hunk 14]
```diff
@@ -520,14 +715,26 @@ public async Task<Entry> VerifiedGetAtTx(byte[] key, ulong tx)
             AtTx = tx
         };
 
-        return await VerifiedGet(keyReq, State());
+        return await VerifiedGet(keyReq, State);
     }
 
+    /// <summary>
+    /// Retrieves with authenticity check the value for a specific key since a transaction ID
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:722]: The meaning of `SinceTx` is somewhat unintuitive. It does not mean to get the change that happened after that transaction but it means that we require the server to index at least that transaction. This is useful when working with asynchronous index processing (noWait=true sent for GRPC mesages).

Thus I'd rephrase the summary to:
```suggestion
    /// Retrieves with authenticity check the value for a specific key. Changes to the key after given transaction can be ignored.
```

File: ImmuDB4Net/ImmuClient.cs [hunk 15]
```diff
@@ -520,14 +715,26 @@ public async Task<Entry> VerifiedGetAtTx(byte[] key, ulong tx)
...
+    /// </summary>
+    /// <param name="key">The lookup key</param>
+    /// <param name="tx">The transaction ID</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:725]:
```suggestion
    /// <param name="tx">The transaction ID after which the server is allowed to ignore changes to the key</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 16]
```diff
@@ -641,6 +891,17 @@ public async Task<List<Entry>> GetAll(List<string> keys)
...
 
+    /// <summary>
+    /// Iterates over the key/values in the selected database and retrieves the values for the matching criteria
+    /// </summary>
+    /// <param name="prefix">Optional, prefix for the keys</param>
+    /// <param name="seekKey">Optional, initial key for the first entry in the iteration</param>
+    /// <param name="endKey">Optional, end key for the scanning range</param>
+    /// <param name="inclusiveSeek">Optional, default is true, specifies if initial key's value is included</param>
+    /// <param name="inclusiveEnd">Optional, default is true, specifies if end key's value is included</param>
+    /// <param name="limit">Optional, maximum number of of returned items</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:902]: I'd also add a remark that if value `0` is used, the default limit will be used (Which is typically 1000). If returned result is larger than the maximum server-side limit, an error is returned.

File: ImmuDB4Net/ImmuClient.cs [hunk 17]
```diff
@@ -641,6 +891,17 @@ public async Task<List<Entry>> GetAll(List<string> keys)
...
     //
 
+    /// <summary>
+    /// Iterates over the key/values in the selected database and retrieves the values for the matching criteria
+    /// </summary>
+    /// <param name="prefix">Optional, prefix for the keys</param>
+    /// <param name="seekKey">Optional, initial key for the first entry in the iteration</param>
+    /// <param name="endKey">Optional, end key for the scanning range</param>
+    /// <param name="inclusiveSeek">Optional, default is true, specifies if initial key's value is included</param>
+    /// <param name="inclusiveEnd">Optional, default is true, specifies if end key's value is included</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:901]:
```suggestion
    /// <param name="inclusiveEnd">Optional, default is false, specifies if end key's value is included</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 18]
```diff
@@ -641,6 +891,17 @@ public async Task<List<Entry>> GetAll(List<string> keys)
     // ========== SCAN ==========
     //
 
+    /// <summary>
+    /// Iterates over the key/values in the selected database and retrieves the values for the matching criteria
+    /// </summary>
+    /// <param name="prefix">Optional, prefix for the keys</param>
+    /// <param name="seekKey">Optional, initial key for the first entry in the iteration</param>
+    /// <param name="endKey">Optional, end key for the scanning range</param>
+    /// <param name="inclusiveSeek">Optional, default is true, specifies if initial key's value is included</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:900]:
```suggestion
    /// <param name="inclusiveSeek">Optional, default is false, specifies if initial key's value is included</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 19]
```diff
@@ -641,6 +891,17 @@ public async Task<List<Entry>> GetAll(List<string> keys)
     // ========== SCAN ==========
     //
 
+    /// <summary>
+    /// Iterates over the key/values in the selected database and retrieves the values for the matching criteria
+    /// </summary>
+    /// <param name="prefix">Optional, prefix for the keys</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:897]: I think this one is mandatory

File: ImmuDB4Net/ImmuClient.cs [hunk 20]
```diff
@@ -904,6 +1290,14 @@ private ImmuState VerifyDualProof(ImmudbProxy.VerifiableTx vtx, Tx tx, ImmuState
     // ========== Z ==========
     //
 
+    /// <summary>
+    /// Adds a scored key to a set
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1294]:
```suggestion
    /// Adds a scored key reference to a sorted set
```

File: ImmuDB4Net/ImmuClient.cs [hunk 21]
```diff
@@ -904,6 +1290,14 @@ private ImmuState VerifyDualProof(ImmudbProxy.VerifiableTx vtx, Tx tx, ImmuState
...
+    /// </summary>
+    /// <param name="set">The set identifier</param>
+    /// <param name="key">The lookup key</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1297]:
```suggestion
    /// <param name="key">The referenced key</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 22]
```diff
@@ -904,6 +1290,14 @@ private ImmuState VerifyDualProof(ImmudbProxy.VerifiableTx vtx, Tx tx, ImmuState
...
+    /// <param name="atTx">The transaction ID</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1298]:
```suggestion
    /// <param name="atTx">Optional, the transaction ID at which the key is bound at</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 23]
```diff
@@ -1003,26 +1420,62 @@ public async Task<TxHeader> VerifiedZAdd(byte[] set, byte[] key, ulong atTx, dou
...
     {
         return await VerifiedZAdd(Utils.ToByteArray(set), Utils.ToByteArray(key), atTx, score);
     }
 
+    /// <summary>
+    /// Iterates over the entries added with ZAdd in the selected database and retrieves the values for the matching criteria
+    /// </summary>
+    /// <param name="set"></param>
+    /// <param name="limit"></param>
+    /// <param name="reverse"></param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1465]:
```suggestion
    /// <param name="set">The set identifier</param>
    /// <param name="limit">Maximum number of entries to return</param>
    /// <param name="reverse">If true, return elements in reversed order</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 24]
```diff
@@ -1262,12 +1764,19 @@ public async Task CreateUser(string user, string password, Iam.Permission permis
         await Service.CreateUserAsync(createUserRequest, Service.GetHeaders(ActiveSession));
     }
 
-    public async Task ChangePassword(string user, string oldPassword, string newPassword)
+    /// <summary>
+    /// Changes a user's password
+    /// </summary>
+    /// <param name="username">The username</param>
+    /// <param name="oldPassword">The username's old password</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1771]:
```suggestion
    /// <param name="oldPassword">The username's old password (only needed when changing sysadmin user's password)</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 25]
```diff
@@ -1279,6 +1788,12 @@ public async Task ChangePassword(string user, string oldPassword, string newPass
     // ========== INDEX MGMT ==========
     //
 
+    /// <summary>
+    /// Flushes the index
+    /// </summary>
+    /// <param name="cleanupPercentage">The cleanup percentage</param>
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1794]:
```suggestion
    /// <param name="cleanupPercentage">The percentage of index data to scan for unused disk space</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 26]
```diff
@@ -1301,11 +1820,27 @@ public async Task CompactIndex()
     // ========== HISTORY ==========
     //
 
+    /// <summary>
+    /// Retrieves the value history of a specific key
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:1824]:
```suggestion
    /// Retrieves the history of changes of a specific key
```
The history can also contain deletions etc.

File: ImmuDB4Net/ImmuClient.cs [hunk 27]
```diff
@@ -1363,4 +1898,156 @@ private List<Tx> buildList(ImmudbProxy.TxList txList)
...
+    {
+        CheckSessionHasBeenOpened();
+        var req = new ImmudbProxy.SQLQueryRequest
+        {
+            Sql = sqlStatement,
+        };
+        if (parameters != null)
+        {
+            int paramNameCounter = 1;
+            foreach (var entry in parameters)
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:2017]: This code is shared between SQLExec and SQLQuery (and will be also used by sync operations) - could be extracted to separate method

File: ImmuDB4Net/ImmuClientSync.cs [hunk 28]
```diff
@@ -16,38 +16,37 @@
```
Comment (byo) [inline ImmuDB4Net/ImmuClientSync.cs:16]: Most of my comments from async client apply here too

File: ImmuDB4Net.Tests/BasicClientIntegrationTests.cs [hunk 29]
```diff
@@ -134,4 +134,45 @@ public async Task Test2()
 
         await client.Close();
     }
+
+     [TestMethod("execute open, verifiedset and verifiedget")]
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net.Tests/BasicClientIntegrationTests.cs:138]: done

File: ImmuDB4Net/Connection.cs [hunk 30]
```diff
@@ -19,22 +19,55 @@
 
 namespace ImmuDB;
 
+/// <summary>
+/// IConnection represents the abstraction for an ImmuClient Connection
+/// </summary>
 public interface IConnection
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/Connection.cs:25]: made it internal

File: ImmuDB4Net/Consts.cs [hunk 31]
```diff
@@ -1,54 +1,58 @@
...
-     * HTree's byte prefix of a node (non-leaf)'s digest.
-     */
+    /// <summary>
+    /// HTree's byte prefix of a node (non-leaf)'s digest. 
+    /// </summary>
     public static readonly byte NODE_PREFIX = 1;
 
-    public static readonly byte SET_KEY_PREFIX = 0;
-    public static readonly byte SORTED_SET_KEY_PREFIX = 1;
+    internal static readonly byte SET_KEY_PREFIX = 0;
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/Consts.cs:21]: you mean to make them public?

File: ImmuDB4Net/KVMetadata.cs [hunk 32]
```diff
@@ -123,16 +142,27 @@ public bool HasExpirationTime
         }
     }
 
-    public long ExpirationTime()
+    /// <summary>
+    /// Gets the expiration time in epoch microsecond
```
Comment (byo) [inline ImmuDB4Net/KVMetadata.cs:146]: This one actually uses seconds precision.

Maybe better to operate on some DateTime structure?

File: ImmuDB4Net/KVMetadata.cs [hunk 33]
```diff
@@ -151,32 +181,62 @@ public byte[] Serialize()
...
 {
+    /// <summary>
+    /// Gets or Sets the expires at value
+    /// </summary>
+    /// <value></value>
     public long ExpiresAt { get; set; }
+    /// <summary>
+    /// Creates an instance of ExpiresAtAttribute
+    /// </summary>
+    /// <param name="expiresAt">The expiration time specified in epoch microseconds</param>
```
Comment (byo) [inline ImmuDB4Net/KVMetadata.cs:221]:
```suggestion
    /// <param name="expiresAt">The expiration time specified in epoch seconds</param>
```

File: ImmuDB4Net/ImmuClient.cs [hunk 34]
```diff
@@ -182,7 +243,62 @@ private void StopHeartbeat()
...
+            {
+                localState = CurrentState;
+                stateHolder.SetState(ActiveSession!, localState);
+            }
+            else
+            {
+                var serverState = Service.CurrentState(new Empty(), Service.GetHeaders(ActiveSession));
+                try
+                {
+                    var verifiableTx = Service.VerifiableTxById(new VerifiableTxRequest
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:261]: it's the local state verified up to server state. They may be well equal. It uses the arguments: SinceTx = localState.TxId, Tx = serverState.TxId,

File: ImmuDB4Net/crypto/InclusionProof.cs [hunk 35]
```diff
@@ -16,18 +16,44 @@
...
+/// </summary>
 public class InclusionProof
 {
+    /// <summary>
+    /// Gets the Leaf ID
+    /// </summary>
+    /// <value></value>
     public int Leaf {get; private set;}
+    /// <summary>
+    /// Gets the width of the inclusion proof
```
Comment (byo) [inline ImmuDB4Net/crypto/InclusionProof.cs:30]:
```suggestion
    /// Gets the width of the Merkle Tree
```

File: ImmuDB4Net/ImmuClient.cs [hunk 36]
```diff
@@ -276,45 +403,59 @@ private void CheckSessionHasBeenOpened()
...
-        ImmudbProxy.ImmutableState state = Service.CurrentState(new Empty(), Service.GetHeaders(ActiveSession));
-        ImmuState immuState = ImmuState.ValueOf(state);
-        if (!immuState.CheckSignature(serverSigningKey))
+        get
         {
-            throw new VerificationException("State signature verification failed");
+            CheckSessionHasBeenOpened();
+            ImmudbProxy.ImmutableState state = Service.CurrentState(new Empty(), Service.GetHeaders(ActiveSession));
+            ImmuState immuState = ImmuState.ValueOf(state);
+            if (!immuState.CheckSignature(serverSigningKey))
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:442]: As the name says, it's the server state. It does not validate against the local state. It is intentional to get the server's current state. Updated comments.

File: ImmuDB4Net/ImmuClient.cs (hunk 19 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:897]: ok, updated

File: ImmuDB4Net/ImmuClient.cs (hunk 18 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:900]: done

File: ImmuDB4Net/ImmuClient.cs (hunk 17 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:901]: fixed

File: ImmuDB4Net/ImmuClient.cs [hunk 37]
```diff
@@ -641,6 +891,17 @@ public async Task<List<Entry>> GetAll(List<string> keys)
...
+    /// <summary>
+    /// Iterates over the key/values in the selected database and retrieves the values for the matching criteria
+    /// </summary>
+    /// <param name="prefix">Optional, prefix for the keys</param>
+    /// <param name="seekKey">Optional, initial key for the first entry in the iteration</param>
+    /// <param name="endKey">Optional, end key for the scanning range</param>
+    /// <param name="inclusiveSeek">Optional, default is true, specifies if initial key's value is included</param>
+    /// <param name="inclusiveEnd">Optional, default is true, specifies if end key's value is included</param>
+    /// <param name="limit">Optional, maximum number of of returned items</param>
+    /// <param name="desc">Optional, specifies the sorting order, defaults to Desc</param>
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:903]: fixed

File: ImmuDB4Net/ImmuClient.cs (hunk 16 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:902]: ok, added this information

File: ImmuDB4Net/ImmuClient.cs [hunk 38]
```diff
@@ -758,12 +1100,19 @@ public async Task<TxHeader> SetAll(List<KVPair> kvList)
         return TxHeader.ValueOf(txHdr);
     }
 
-    public async Task<TxHeader> SetReference(byte[] key, byte[] referencedKey, ulong atTx)
+    /// <summary>
+    /// Adds a tag (reference) to a specific key/value element in the selected database
+    /// </summary>
+    /// <param name="reference">The reference</param>
+    /// <param name="referencedKey">The lookup key</param>
+    /// <param name="atTx">Transaction ID</param>
```
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1108]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 6 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:173]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 8 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:185]: This can use the default name 'defaultdb'. The aim was to make the use of this SDK as simple as possible, this was stated by our leadership.

File: ImmuDB4Net/ImmuClient.cs (hunk 10 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:338]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 9 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:340]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 11 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:407]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 12 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:488]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 13 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:513]: Renamed. Most likely 'borrowed' from Java SDK

File: ImmuDB4Net/ImmuClient.cs (hunk 14 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:722]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 15 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:725]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 20 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1294]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 22 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1298]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 23 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1465]: ok, missed that

File: ImmuDB4Net/ImmuClient.cs (hunk 24 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1771]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 25 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1794]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 26 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1824]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 27 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:2017]: ok, refactored, created in sql the Converters class

File: ImmuDB4Net/ImmuClient.cs (hunk 5 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:183]: ok

File: ImmuDB4Net/ImmuClient.cs (hunk 21 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:1297]: ok

File: ImmuDB4Net/Consts.cs (hunk 31 above)
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/Consts.cs]: I've changed the visibility and added comments

File: ImmuDB4Net/ImmuClient.cs (hunk 34 above)
[the remark this answers is not in the document]
Comment (byo) [reply ImmuDB4Net/ImmuClient.cs:261]: Sure, but I'm asking explicitly about the Service.VerifiableTxById call - if I understand correctly, this is automatic code for GRPC communitation so it does not check the proof on the client side.

I also don't see any code that would check if the `localState` retrieved in line 250 is checked against the response from the server (we only give the TxId on line 263). I was wondering where the consistency proof is executed.

File: ImmuDB4Net/ImmuClient.cs (hunk 8 above)
Comment (byo) [reply ImmuDB4Net/ImmuClient.cs:185]: I can see that this could be a shorthand version, but I still can't find the use-case for it.

When opening a connection first the constructor is callde - so here we would provide the db name.
But then we have to call Open() again which requires db name. So why we ever need to specify the db name in constructor?

It would make sense to me if there was some alternative `Open` method that does not take the db name parameter and instead takes the one set in constructor but I didn't see one.

File: ImmuDB4Net/ImmuClient.cs [hunk 39]
```diff
@@ -206,7 +322,8 @@ public async Task Open(string username, string password, string defaultdb)
                 Address = GrpcAddress,
                 ShutdownTimeout = ConnectionShutdownTimeout
             });
-            ActiveSession = await SessionManager.OpenSessionAsync(Connection, username, password, defaultdb);
+            ActiveSession = await SessionManager.OpenSessionAsync(Connection, username, password, databaseName);
+            ValidateLocalState();
             heartbeatCloseRequested = new ManualResetEvent(false);
```
Comment (byo) [inline ImmuDB4Net/ImmuClient.cs:327]: we don't set the value of `currentDb` here.

File: ImmuDB4Net/ImmuClient.cs (hunk 8 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:185]: Removed this overload.

File: ImmuDB4Net/KVMetadata.cs (hunk 33 above)
Comment (raducrisan1) [reply ImmuDB4Net/KVMetadata.cs:221]: done

File: ImmuDB4Net/KVMetadata.cs (hunk 32 above)
Comment (raducrisan1) [reply ImmuDB4Net/KVMetadata.cs:146]: changed to DateTime

File: ImmuDB4Net/crypto/InclusionProof.cs (hunk 35 above)
Comment (raducrisan1) [reply ImmuDB4Net/crypto/InclusionProof.cs:30]: updated

File: ImmuDB4Net/ImmuClient.cs (hunk 39 above)
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:327]: oops, fixed.

File: ImmuDB4Net/KVMetadata.cs (hunk 32 above)
Comment (raducrisan1) [reply ImmuDB4Net/KVMetadata.cs:146]: changed to DateTime

File: ImmuDB4Net/ImmuClient.cs (hunk 34 above)
[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:261]: We should discuss in a call what is expected here. Initially I followed what Jeronimo suggested.

[the remark this answers is not in the document]
Comment (raducrisan1) [reply ImmuDB4Net/ImmuClient.cs:250]: I've added the dualproof verification on both `ImmuClient` and `ImmuClientSync`

File: ImmuDB4Net/KVMetadata.cs (hunk 33 above)
Comment (byo) [reply ImmuDB4Net/KVMetadata.cs]: Actually I think it would be good here to accept the DateTime parameter - WDYT?
Of course documentation would have to remark that it will be truncated to second precision

**Review by byo — APPROVED** [review]

File: ImmuDB4Net/KVMetadata.cs (hunk 33 above)
Comment (raducrisan1) [reply ImmuDB4Net/KVMetadata.cs]: done, added an overloaded constructor
